### PR TITLE
PoS Decouple Validator Jailing from Validator Snapshotting

### PR DESCRIPTION
**Pull request by tholonious**

### CODE REVIEW

**Comment by tholonious** [comment]
Current dependencies on/for this PR:
* feature/proof-of-stake
  * **PR #579** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/579" target="_blank">Graphite</a> 
    * **PR #581** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/581" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/deso-protocol/core/581?utm_source=stack-comment).

File: lib/block_view_validator.go [hunk 1]
```diff
@@ -1937,6 +1937,65 @@ func (bav *UtxoView) GetGlobalActiveStakeAmountNanos() (*uint256.Int, error) {
...
+		if shouldJailValidator {
+			if err = bav.JailValidator(validatorEntry); err != nil {
+				return errors.Wrapf(
+					err, "JailAllInactiveValidators: problem jailing validator %v: ", validatorEntry.ValidatorPKID,
+				)
+			}
+		}
+
+		// We don't want to retrieve any ValidatorEntries from the db that are present in the UtxoView.
+		utxoViewValidatorPKIDs = append(utxoViewValidatorPKIDs, validatorEntry.ValidatorPKID)
```
Comment (lazynina) [inline lib/block_view_validator.go:1969]: I think we want to add all validator PKIDs to the utxo validator PKIDs so we don't bother fetching them from the DB even if the validator entry is deleted? also we can reduce the nested if.
```suggestion
		// We don't want to retrieve any ValidatorEntries from the db that are present in the UtxoView.
		utxoViewValidatorPKIDs = append(utxoViewValidatorPKIDs, validatorEntry.ValidatorPKID)
		
		if validatorEntry.isDeleted {
			continue
		}

		// Check if we should jail the validator.
		shouldJailValidator, err := bav.ShouldJailValidator(validatorEntry, blockHeight)
		if err != nil {
			return errors.Wrapf(
				err,
				"JailAllInactiveValidators: problem determining if should jail validator %v: ",
				validatorEntry.ValidatorPKID,
			)
		}

		// Jail them if so.
		if !shouldJailValidator {
			continue
		}
		
		if err = bav.JailValidator(validatorEntry); err != nil {
			return errors.Wrapf(
				err, "JailAllInactiveValidators: problem jailing validator %v: ", validatorEntry.ValidatorPKID,
			)
		}
```

File: lib/block_view_validator.go [hunk 2]
```diff
@@ -1937,6 +1937,65 @@ func (bav *UtxoView) GetGlobalActiveStakeAmountNanos() (*uint256.Int, error) {
...
+		// Check if we should jail the validator.
+		shouldJailValidator, err := bav.ShouldJailValidator(validatorEntry, blockHeight)
+		if err != nil {
+			return errors.Wrapf(
+				err, "JailAllInactiveValidators: problem determining if should jail validator %v: ", validatorEntry.ValidatorPKID,
+			)
+		}
+		// Jail them if so.
+		if shouldJailValidator {
+			if err = bav.JailValidator(validatorEntry); err != nil {
```
Comment (lazynina) [inline lib/block_view_validator.go:1988]: reduce nested if? up to you really, but I prefer to `continue` instead of the nested if err pattern

**Review by lazynina — CHANGES_REQUESTED** [review]

File: lib/block_view_validator.go (hunk 1 above)
Comment (tholonious) [reply lib/block_view_validator.go:1969]: Great catch. Thanks!

File: lib/block_view_validator.go (hunk 2 above)
Comment (tholonious) [reply lib/block_view_validator.go:1988]: Agreed. Good suggestion

File: lib/block_view_validator.go [hunk 3]
```diff
@@ -1937,6 +1937,69 @@ func (bav *UtxoView) GetGlobalActiveStakeAmountNanos() (*uint256.Int, error) {
...
+		shouldJailValidator, err := bav.ShouldJailValidator(validatorEntry, blockHeight)
+		if err != nil {
+			return errors.Wrapf(
+				err,
+				"JailAllInactiveValidators: problem determining if should jail validator %v: ",
+				validatorEntry.ValidatorPKID,
+			)
+		}
+
+		// Jail them if so.
```
Comment (lazynina) [inline lib/block_view_validator.go:1962]:
```suggestion
		// If this validator should not be jailed, continue to the next validator.
```

File: lib/block_view_validator.go [hunk 4]
```diff
@@ -1937,6 +1937,69 @@ func (bav *UtxoView) GetGlobalActiveStakeAmountNanos() (*uint256.Int, error) {
...
+		if !shouldJailValidator {
+			continue
+		}
+
```
Comment (lazynina) [inline lib/block_view_validator.go:1966]:
```suggestion
// If we're here, this validator should be jailed.
```

or something like that.

File: lib/block_view_validator.go [hunk 5]
```diff
@@ -1937,6 +1937,69 @@ func (bav *UtxoView) GetGlobalActiveStakeAmountNanos() (*uint256.Int, error) {
...
+	for _, validatorEntry := range dbValidatorEntries {
+		// Check if we should jail the validator.
+		shouldJailValidator, err := bav.ShouldJailValidator(validatorEntry, blockHeight)
+		if err != nil {
+			return errors.Wrapf(
+				err, "JailAllInactiveValidators: problem determining if should jail validator %v: ", validatorEntry.ValidatorPKID,
+			)
+		}
+		// Jail them if so.
```
Comment (lazynina) [inline lib/block_view_validator.go:1988]: comment clean up

**Review by lazynina — CHANGES_REQUESTED** [review]

**Review by lazynina — APPROVED** [review]

**Pull request merged**